src/map/clif.c

```diff
@@ -19076,6 +19100,147 @@ const char *clif_get_bl_name(const struct block_list *bl)
+
+	length = 8 + 2 * NAME_LENGTH + MAP_NAME_LENGTH_EXT + 2;
+	fd = sd->fd;
+
+	WFIFOHEAD(fd, length);
+	memset(WFIFOP(fd, 0), 0, length);
+	WFIFOW(fd, 0) = 0x98a;
+	WFIFOL(fd, 4) = c->clan_id;
+	offset = 8;
+	memcpy(WFIFOP(fd, offset), c->name, NAME_LENGTH);
+	offset += NAME_LENGTH;
+	memcpy(WFIFOP(fd, offset), c->master, NAME_LENGTH);
+	offset += NAME_LENGTH;
+	mapindex->getmapname_ext(c->map, mapname);
+	safestrncpy(WFIFOP(fd, offset), mapname, MAP_NAME_LENGTH_EXT);
+	offset += MAP_NAME_LENGTH_EXT;
+
+	WFIFOB(fd, offset++) = clan->ally_count(c);
+	WFIFOB(fd, offset++) = clan->antagonist_count(c);
+
+	for (i = 0; i < clan->maxalliances; i++) {
+		if (c->allies[i].clan_id > 0) {
+			ally = clan->search(c->allies[i].clan_id);
+			if (ally != NULL) {
+				safestrncpy(WFIFOP(fd, offset), ally->name, NAME_LENGTH);
+				offset += NAME_LENGTH;
+			}
+		}
+	}
+
+	for (i = 0; i < clan->maxalliances; i++) {
+		if (c->antagonists[i].clan_id > 0) {
+			antagonist = clan->search(c->antagonists[i].clan_id);
+			if (antagonist != NULL) {
+				safestrncpy(WFIFOP(fd, offset), antagonist->name, NAME_LENGTH);
+				offset += NAME_LENGTH;
+			}
+		}
+	}
+
+	WFIFOW(fd, 2) = offset;
+	WFIFOSET(fd, offset);
+#endif
+}
+
+/**
+ * Clan System: Updates the online and maximum player count of a clan.
+ * 0988 <online count>.W <maximum member amount>.W (ZC_NOTIFY_CLAN_CONNECTINFO)
+ */
+void clif_clan_onlinecount(struct clan *c)
+{
+#if PACKETVER >= 20120716
+	uint8 buf[6];
+	struct map_session_data *sd;
+
+	WBUFW(buf, 0) = 0x988;
+	WBUFW(buf, 2) = c->connect_member;
+	WBUFW(buf, 4) = c->max_member;
+
+	if ((sd = clan->getonlinesd(c)) != NULL)
+	  clif->send(buf, 6, &sd->bl, CLAN);
+#endif
+}
+
+/**
+* Clan System: Notifies the client that the player has left his clan.
+* 0989 (ZC_ACK_CLAN_LEAVE)
+**/
+void clif_clan_leave(struct map_session_data* sd)
+{
+#if PACKETVER >= 20131223
+	int fd;
+	
+	nullpo_retv (sd);
+	
+	fd = sd->fd;
+
+	WFIFOHEAD(fd, 2);
+	WFIFOW(fd, 0) = 0x989;
+	WFIFOSET(fd, 2);
+#endif
+}
+
+/**
+ * Clan System: Sends a clan message to a player
+ * 098e <length>.W <name>.24B <message>.?B (ZC_NOTIFY_CLAN_CHAT)
+ */
+void clif_clan_message(struct clan *c, const char *mes, int len)
+{
+#if PACKETVER >= 20120716
+	struct map_session_data *sd;
+	uint8 buf[256];
+
+	if (len == 0) {
+		return;
+	} else if (len > (sizeof(buf) - 5 - NAME_LENGTH)) {
+		ShowWarning("clif_clan_message: Truncated message '%s' (len=%d, max=%"PRIuS", clan_id=%d).\n", mes, len, sizeof(buf) - 5, c->clan_id);
+		len = sizeof(buf) - 5 - NAME_LENGTH;
+	}
+
+	WBUFW(buf, 0) = 0x98e;
+	WBUFW(buf, 2) = len + 5 + NAME_LENGTH;
+	
+	safestrncpy(WBUFP(buf, 4 + NAME_LENGTH), mes, len + 1);
+
+	if ((sd = clan->getonlinesd(c)) != NULL)
+	  clif->send(buf, WBUFW(buf, 2), &sd->bl, CLAN);
+#endif
+}
+
+/**
+ * Clan System: Parses a clan message from a player.
+ * 098d <length>.W <text>.?B (<name> : <message>) (CZ_CLAN_CHAT)
+ */
+void clif_parse_ClanMessage(int fd, struct map_session_data* sd)
```

Comment:
This should have a forward-declared `nonull` attribute check like other clif functions to make sure that the socket-invoked function didn't send a null sd.